src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DevTools/RemoteDevToolsService.cs

```diff
@@ -0,0 +1,159 @@
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var contents = codeDocument.GetCSharpSourceText().ToString();
+        var filePath = documentContext.Snapshot.FilePath + ".g.cs";
+        
+        return new Microsoft.CodeAnalysis.Razor.Protocol.DevTools.DocumentContentsResponse
+        {
+            Contents = contents,
+            FilePath = filePath
+        };
+    }
+
+    private async ValueTask<Microsoft.CodeAnalysis.Razor.Protocol.DevTools.DocumentContentsResponse?> GetHtmlDocumentTextAsync(RemoteDocumentContext documentContext, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var contents = codeDocument.GetHtmlSourceText().ToString();
+        var filePath = documentContext.Snapshot.FilePath + ".g.html";
+        
+        return new Microsoft.CodeAnalysis.Razor.Protocol.DevTools.DocumentContentsResponse
+        {
+            Contents = contents,
+            FilePath = filePath
+        };
+    }
+
+    private async ValueTask<Microsoft.CodeAnalysis.Razor.Protocol.DevTools.DocumentContentsResponse?> GetFormattingDocumentTextAsync(RemoteDocumentContext documentContext, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+#pragma warning disable CS0618 // Type or member is obsolete
+        var contents = CSharpFormattingPass.GetFormattingDocumentContentsForSyntaxVisualizer(codeDocument);
+#pragma warning restore CS0618 // Type or member is obsolete
+        var filePath = documentContext.Snapshot.FilePath + ".formatting.cs";
+        
+        return new Microsoft.CodeAnalysis.Razor.Protocol.DevTools.DocumentContentsResponse
+        {
+            Contents = contents,
+            FilePath = filePath
+        };
+    }
+
+    private async ValueTask<string> GetTagHelpersJsonAsync(RemoteDocumentContext documentContext, Microsoft.CodeAnalysis.Razor.Protocol.DevTools.TagHelpersKind kind, CancellationToken cancellationToken)
+    {
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var tagHelpers = kind switch
+        {
+            Microsoft.CodeAnalysis.Razor.Protocol.DevTools.TagHelpersKind.All => codeDocument.GetTagHelpers(),
+            Microsoft.CodeAnalysis.Razor.Protocol.DevTools.TagHelpersKind.InScope => codeDocument.GetRequiredTagHelperContext().TagHelpers,
+            Microsoft.CodeAnalysis.Razor.Protocol.DevTools.TagHelpersKind.Referenced => (IEnumerable<TagHelperDescriptor>?)codeDocument.GetReferencedTagHelpers(),
+            _ => []
+        };
+
+        tagHelpers ??= [];
+        return JsonSerializer.Serialize(tagHelpers, new JsonSerializerOptions { WriteIndented = true });
```

Comment:
Rather than a string, this should actually return an ImmutableArray<TagHelperDescriptor>, and just return the tagHelpers variable. Serialization can happen in the SyntaxVisualizerControl